File: kotlin-sdk/src/main/kotlin/com/cleanster/api/WebhooksApi.kt
package com.cleanster.api

import com.cleanster.CleansterClient
import com.cleanster.model.*

/** API methods for registering and managing webhook endpoints. */
class WebhooksApi internal constructor(private val client: CleansterClient) {

    /** List all registered webhook endpoints. */
    suspend fun listWebhooks(): ApiResponse<List<Any>> = client.request(
        method = "GET",
        path   = "/v1/webhooks",
    )

    /**
     * Register a new webhook endpoint.
     *
     * @param url   HTTPS URL that will receive POST payloads.
     * @param event Event name to subscribe to (e.g. `"booking.completed"`).
     */
    suspend fun createWebhook(url: String, event: String): ApiResponse<Webhook> = client.request(
        method = "POST",
        path   = "/v1/webhooks",
        body   = CreateWebhookRequest(url = url, event = event),
    )

    /** Update an existing webhook's URL or event. */
    suspend fun updateWebhook(webhookId: Int, url: String, event: String): ApiResponse<Webhook> = client.request(
        method = "PUT",
        path   = "/v1/webhooks/$webhookId",
        body   = CreateWebhookRequest(url = url, event = event),
    )

    /** Delete a webhook endpoint. */
    suspend fun deleteWebhook(webhookId: Int): ApiResponse<Map<String, Any>> = client.request(
        method = "DELETE",
        path   = "/v1/webhooks/$webhookId",
    )
}

File: swift-sdk/Sources/Cleanster/Api/WebhooksApi.swift
import Foundation

/// API methods for registering and managing webhook endpoints.
public final class WebhooksApi {
    private let client: CleansterClient
    init(client: CleansterClient) { self.client = client }

    /// List all registered webhook endpoints.
    public func listWebhooks() async throws -> ApiResponse<AnyCodable> {
        return try await client.requestRaw(method: "GET", path: "/v1/webhooks")
    }

    /// Register a new webhook endpoint.
    ///
    /// - Parameters:
    ///   - url: HTTPS URL that will receive the POST payload.
    ///   - event: The event name to subscribe to (e.g. `"booking.completed"`).
    public func createWebhook(url: String, event: String) async throws -> ApiResponse<Webhook> {
        let body = CreateWebhookRequest(url: url, event: event)
        return try await client.request(method: "POST", path: "/v1/webhooks", body: body)
    }

    /// Update an existing webhook's URL or event.
    public func updateWebhook(_ webhookId: Int, url: String, event: String) async throws -> ApiResponse<Webhook> {
        let body = CreateWebhookRequest(url: url, event: event)
        return try await client.request(method: "PUT", path: "/v1/webhooks/\(webhookId)", body: body)
    }

    /// Delete a webhook endpoint.
    public func deleteWebhook(_ webhookId: Int) async throws -> ApiResponse<AnyCodable> {
        return try await client.requestRaw(method: "DELETE", path: "/v1/webhooks/\(webhookId)")
    }
}
